src/main/kotlin/graphics/scenery/Arrow.kt

```diff
@@ -0,0 +1,168 @@
+package graphics.scenery
+
+import cleargl.GLVector
+import graphics.scenery.backends.ShaderType
+import java.nio.FloatBuffer
+import java.nio.IntBuffer
+
+/**
+ * Class for creating 3D arrows, derived from [Node] and using [HasGeometry].
+ * The arrow is defined with [vector] having the vector's foot/start as the
+ * reference coordinate (arrow's foot will appear at the coordinate given to setPosition()).
+ *
+ *   /|\
+ *  / | \
+ * /--+--\
+ *    |
+ *    |
+ *    |
+ * The arrow is created as the main segment (drawn vertically bottom to up),
+ * then 3 segments to build a triangle and then another triangle perpendicular
+ * to the former one; altogehter 7 segments drawn sequentially.
+ * The arrow head scales with the length of the arrow.
+ *
+ * @author Vladimir Ulman <ulman@mpi-cbg.de>
+ */
+class Arrow(var vector: GLVector) : Node("Arrow"), HasGeometry {
+    /** Size of one vertex (e.g. 3 in 3D) */
+    override val vertexSize: Int = 3
+    /** Size of one texcoord (e.g. 2 in 3D) */
+    override val texcoordSize: Int = 2
+    /** Geometry type -- Default for Line is [GeometryType.LINE] */
+    override var geometryType: GeometryType = GeometryType.LINE_STRIP_ADJACENCY
+    /** Vertex buffer */
+    override var vertices: FloatBuffer = BufferUtils.allocateFloat(30)
+    /** Normal buffer */
+    override var normals: FloatBuffer = BufferUtils.allocateFloat(30)
+    /** Texcoord buffer */
+    override var texcoords: FloatBuffer = BufferUtils.allocateFloat(20)
+    /** Index buffer */
+    override var indices: IntBuffer = IntBuffer.wrap(intArrayOf())
+
+    /** Shader property for the line's starting segment color. Consumed by the renderer. */
+    @ShaderProperty
+    var startColor = GLVector(0.0f, 1.0f, 0.0f, 1.0f)
+
+    /** Shader property for the line's color. Consumed by the renderer. */
+    @ShaderProperty
+    var lineColor = GLVector(1.0f, 1.0f, 1.0f, 1.0f)
+
+    /** Shader property for the line's end segment color. Consumed by the renderer. */
+    @ShaderProperty
+    var endColor = GLVector(0.7f, 0.5f, 0.5f, 1.0f)
+
+    /** Shader property for the line's cap length (start and end caps). Consumed by the renderer. */
+    @ShaderProperty
+    var capLength = 1
+
+    /** Shader property to keep track of the current number of vertices. Consumed by the renderer. */
+    @ShaderProperty
+    var vertexCount: Int = 0
+        private set
+
+    /** Shader property for the line's edge width. Consumed by the renderer. */
+    @ShaderProperty
+    var edgeWidth = 2.0f
+
+    init {
+		material = ShaderMaterial.fromClass(this::class.java, listOf(ShaderType.VertexShader, ShaderType.GeometryShader, ShaderType.FragmentShader))
```

Comment:
You are declaring a `ShaderMaterial` that references the current class, and uses {vertex, geometry, fragment} shaders. Where are the shader files?

Reply:
So what this line does is instruct the renderer to look for files named `$classSimpleName.vert/.frag/.geom` when a Node of this type is initialized. What will happen here is that the renderer will bail out because it can't find the shaders named `Arrow.vert` and so on. The way to fix this is to write `ShaderMaterial.fromClass(Line::class.java,...)`, which will then use the shaders from Line, which is a good idea imho :)

Reply:
aha! changed here 2447569de5c0ddf4d892121fe47e49ddb48189c2